B-22044-INT-b

stevengleason-caci opened the pull request:
## [B-22044](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A1040112)

## Summary
Second INT pull request. Fix [I-14262](https://www13.v1host.com/USTRANSCOM38/Issue.mvc/Summary?oidToken=Issue%3A1114155), capture the office user's new OktaID when their request is approved and their Okta account is created.

If there are any errors during this new step, they are only logged, they shouldn't interrupt the existing approval process.

## Verification Steps for the Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [ ] Have the Agility acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [ ] Has the branch been pulled in and checked out?
- [ ] Have the BL acceptance criteria been met for this change?
- [ ] Was the CircleCI build successful?
- [ ] Has the code been reviewed from a standards and best practices point of view?

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test
If you have Okta-test admin access, you can see that the account is created and deleted.

1. Enable Okta account creation for local dev. In `requested_office_users.go` line 360, remove or modify the "devlocal" check.
2. Go to http://officelocal:3000/sign-in and request an account.
3. Log in to the local admin app, Requested Office Users, and approve the request.
4. Log in to test-milmove.okta.mil admin, directory, and see that the new account was created.
5. In the local admin app, Office Users, edit and delete the new user.
6. In Okta, see that the new account is gone.

### Frontend

- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://transcom.github.io/mymove-docs/docs/adrs/Browser-Support/#minimum-browser-requirements) (Chrome, Firefox, Edge).
- [ ] There are no new console errors in the browser devtools.
- [ ] There are no new console errors in the test output.
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.
- [ ] This change meets the standards for [Section 508 compliance](https://www.ssa.gov/accessibility/andi/help/install.html).

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/getting-started/development/logging).
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/backend/guides/golang-guide#querying-the-database-safely) have been satisfied.

### Database

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#zero-downtime-migrations).
- [ ] Have been communicated to #g-database.
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#secure-migrations).

## Screenshots

ryan-mchugh commented on pkg/handlers/adminapi/requested_office_users_test.go:
```diff
@@ -952,15 +1084,331 @@ func (suite *HandlerSuite) TestUpdateRequestedOfficeUserHandlerWithOktaAccountCr
+		suite.Equal(originalOktaID, updatedUser.OktaID)
+	})
+
+	suite.Run("Overwrite an existing OktaID", func() {
```
Does this really "overwrite" an existing OktaID since it's carrying on the OktaID that's already present `idFromOkta`.

stevengleason-caci replied:
Yes. For this test I set up the user with BuildDefaultUser which generates a random OktaID. I assert that on lines 1303 and 1304. That value in our Users table gets updated to be the value from the (mock) Okta api call, idFromOkta.

I based this behavior on what I found in pkg/handlers/authentication/auth.go, line 1330, where the OktaID is updated if it doesn't match what we already had.

stevengleason-caci commented:
> One of the linters is failing in gitlab. I'll make this a draft while I look into that.

I added some upstream dependencies to apply those changes to my new tests: replaced RoleAssociator with RoleFetcher, and added the UserPrivilegeAssociator.

Ready to review again.

brianmanley-caci reviewed (approved):
Working well for me ✅

loganwc reviewed (approved):
looks good, nice tests!

stevengleason-caci commented on pkg/handlers/adminapi/requested_office_users_test.go:
```diff
@@ -1457,15 +1467,461 @@ func (suite *HandlerSuite) TestUpdateRequestedOfficeUserHandlerWithOktaAccountCr
+		userPrivilegeAssociator := &mocks.UserPrivilegeAssociator{}
+		requestedOfficeUserUpdater := &mocks.RequestedOfficeUserUpdater{}
+
+		status := "APPROVED"
```
I need to make some updates anyway. I can do this too.

stevengleason-caci commented:
More changes in integrationTesting broke this again. I'll make this a draft while I fix it.
